File: solutions/binary-tree-level-order-traversal/binary-tree-level-order-traversal.cpp
// @Title: Binary Tree Level Order Traversal
// @Author: colinjxc
// @Date: 2022-01-26T13:20:19+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-level-order-traversal


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        queue<TreeNode*> _q;
        _q.push(root);

        vector<int> tmp;
        while (!_q.empty()) {
            int n = _q.size();
            for (int i = 0; i < n; i++) {
                TreeNode* node = _q.front();
                _q.pop();
                if (node == nullptr) continue;
                tmp.push_back(node->val);
                _q.push(node->left);
                _q.push(node->right);
            }
            if (tmp.size() > 0) {
                res.push_back(tmp);
                tmp.clear();
            }
        }
        return res;
    }
};

File: solutions/binary-tree-level-order-traversal/binary-tree-level-order-traversal.go
// @Title: Binary Tree Level Order Traversal
// @Author: colinjxc
// @Date: 2022-02-17T00:22:48+08:00
// @URL: https://leetcode-cn.com/problems/binary-tree-level-order-traversal


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func levelOrder(root *TreeNode) [][]int {
    res := make([][]int, 0)
    q := make([]*TreeNode, 0)

    q = append(q, root)
    for len(q) > 0 {
        tmp := []int{}
        size := len(q)
        for i := 0; i < size; i++ {
            node := q[i]
            if node != nil {
                tmp = append(tmp, node.Val)
                q = append(q, node.Left, node.Right)
            }
        }
        q = q[size:]
        if len(tmp) > 0 {
            res = append(res, tmp)
        }
    }
    return res;
}
